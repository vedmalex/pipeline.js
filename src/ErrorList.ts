export class ErrorList extends Error {
  errors: Array<any>
  constructor(_list: Array<any> | any) {
    super('Complex Error')
    if (Array.isArray(_list)) {
      const list = _list.filter(e => e)
      if (list.length > 1) {
        this.errors = Array.isArray(list) ? list : [list]
      } else if ((list.length = 1)) {
        return list[0]
      } else {
        return null
      }
    } else {
      return _list
    }
  }
}
