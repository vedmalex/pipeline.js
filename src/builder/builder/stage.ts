import { Stage } from '../../stage/stage';
import { StageConfig } from '../../stage/StageConfig';
import { UnsetMarker } from '../../stage/types';
import { BuilderDef, StageBuilder } from './../types';


export function stage<TConfig extends StageConfig<any, any>>(
  _def: Partial<BuilderDef<TConfig>> = {}
): StageBuilder<{
  _type: UnsetMarker;
  _input: UnsetMarker;
  _output: UnsetMarker;
  _usage: {
    output: 1;
    run: 1;
    build: 1;
  };
  _run: UnsetMarker;
  _stage: UnsetMarker;
}> {
  return {
    _def: _def as BuilderDef<TConfig>,
    input(input) {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      _def.cfg.input = input;
      return stage({
        ..._def,
        inputs: input,
      }) as any;
    },
    output(output) {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      _def.cfg.output = output;
      return stage({
        ..._def,
        outputs: output,
      }) as any;
    },
    run(run) {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      _def.cfg.run = run as any;
      return stage({
        ..._def,
      }) as any;
    },
    build() {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      return new Stage(_def.cfg) as any;
    },
  };
}
