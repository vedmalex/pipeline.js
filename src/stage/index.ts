export * from './Context'
export * from './errors'
export * from './getStageConfig'
export * from './stage'
export * from './StageConfig'
export * from './types'
export * from './utils'
