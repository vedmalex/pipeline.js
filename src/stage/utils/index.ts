export * from './empty_run'
export * from './execute_callback'
export * from './execute_custom_run'
export * from './execute_rescue'
export * from './execute_validate'
export * from './run_callback_once'
export * from './run_or_execute'
