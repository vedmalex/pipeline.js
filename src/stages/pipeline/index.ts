export * from './pipeline'
export * from './PipelineConfig'
