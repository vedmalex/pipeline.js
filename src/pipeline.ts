import { CreateError } from './utils/ErrorList'
import { run_or_execute } from './utils/run_or_execute'
import { Stage } from './stage'
import { empty_run } from './utils/empty_run'
import {
  isIStage,
  IStage,
  CallbackFunction,
  PipelineConfig,
  StageRun,
  StageConfig,
  RunPipelineFunction,
} from './utils/types'

/**
 * it make possible to choose which stage to run according to result of `condition` evaluation
 *  - config as
 		- `Function` --- first Stage for pipeline
 * 		- `Object` --- config for Pipeline
 *			  - `stages` list of stages
 *			  - `name` name of pipeline
 * 		- `Empty` --- empty pipeline
 *
 * @param {Object} config configuration object
 */
export class Pipeline<T, C extends PipelineConfig<T, R>, R> extends Stage<
  T,
  C,
  R
> {
  stages!: Array<IStage<any, any, any> | RunPipelineFunction<any, any>>

  constructor(config?: string | PipelineConfig<T, R>) {
    let stages: Array<IStage<any, any, any> | RunPipelineFunction<any, any>> =
      []
    if (typeof config == 'object') {
      if (config instanceof Stage) {
        stages.push(config)
        super()
      } else {
        if (config.run instanceof Function) {
          stages.push(config.run)
          delete config.run
        }
        if (Array.isArray(config.stages)) {
          stages.push(...config.stages)
        }
        if (config instanceof Array) {
          stages.push.apply(stages, config)
        }
        super(config as C)
      }
      this.stages = []
      for (let i = 0; i < stages.length; i++) {
        this.addStage(stages[i])
      }
    } else if (typeof config == 'string') {
      super(config)
      this.stages = []
    } else if (config) {
      throw CreateError('wrong arguments check documentation')
    } else {
      super()
      this.stages = []
    }
  }

  public override get reportName() {
    return `PIPE:${this.config.name ? this.config.name : ''}`
  }

  addStage(
    _stage:
      | StageConfig<T, R>
      | RunPipelineFunction<any, any>
      | IStage<any, any, any>,
  ) {
    let stage: IStage<any, any, any> | RunPipelineFunction<any, any> | undefined
    if (typeof _stage === 'function') {
      stage = _stage
    } else {
      if (typeof _stage === 'object') {
        if (!isIStage(_stage)) {
          stage = new Stage(_stage)
        } else {
          stage = _stage
        }
      }
    }
    if (stage) {
      this.stages.push(stage)
      this.run = undefined
    }
  }

  public override toString() {
    return '[pipeline Pipeline]'
  }

  override compile(rebuild: boolean = false): StageRun<T, R> {
    let run: StageRun<T, R> = (
      err: Error | undefined,
      context: T | R,
      done: CallbackFunction<T | R>,
    ) => {
      let i = -1
      //sequential run;
      let next = (err: Error | undefined, context: T | R) => {
        i += 1
        if (i < this.stages.length) {
          run_or_execute(this.stages[i], err, context, next)
        } else if (i == this.stages.length) {
          done(err, context)
        } else {
          done(new Error('done call more than once'), context)
        }
      }
      next(err, context)
    }

    if (this.stages.length > 0) {
      this.run = run
    } else {
      this.run = empty_run
    }

    return super.compile(rebuild)
  }
}
