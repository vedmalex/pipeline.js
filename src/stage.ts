import { CreateError } from './utils/ErrorList'
import {
  CallbackFunction,
  EnsureFunction,
  IStage,
  RunPipelineFunction,
  StageConfig,
  StageRun,
  ValidateFunction,
} from './utils/types'

import Ajv from 'ajv'

import ajvFormats from 'ajv-formats'
import ajvKeywords from 'ajv-keywords'
import ajvErrors from 'ajv-errors'
import { execute_ensure } from './utils/execute_ensure'
import { execute_validate } from './utils/execute_validate'
import { execute_rescue } from './utils/execute_rescue'
import { execute_callback } from './utils/execute_callback'
import { can_fix_error } from './utils/can_fix_error'

// make possibility to context be immutable for debug purposes

export class Stage<T = any, C extends StageConfig<T, R> = any, R = T>
  implements IStage<T, C, R>
{
  public get config(): C {
    return this._config
  }
  protected _config: C
  constructor(config?: string | C | RunPipelineFunction<T, R>) {
    this._config = {} as C
    if (typeof config == 'string') {
      this._config.name = config
    } else if (typeof config == 'function') {
      this._config.run = config
    } else if (typeof config == 'object') {
      if (config.name) {
        this._config.name = config.name
      }
      if (config.rescue) {
        this._config.rescue = config.rescue
      }
      if (config.run) {
        this._config.run = config.run
      }
      if (config.validate && config.schema) {
        throw CreateError('use only one `validate` or `schema`')
      }
      if (config.ensure && config.schema) {
        throw CreateError('use only one `ensure` or `schema`')
      }
      if (config.ensure && config.validate) {
        throw CreateError('use only one `ensure` or `validate`')
      }
      if (config.validate) {
        this._config.validate = config.validate
      }
      if (config.ensure) {
        this._config.ensure = config.ensure
      }
      if (config.compile) {
        this._config.compile = config.compile
      }
      if (config.precompile) {
        this._config.precompile = config.precompile
      }
      if (config.schema) {
        this._config.schema = config.schema
        const ajv = new Ajv({ allErrors: true })
        ajvFormats(ajv)
        ajvErrors(ajv, { singleError: true })
        ajvKeywords(ajv)
        const validate = ajv.compile(config.schema)
        this._config.validate = (ctx: T): boolean => {
          if (!validate(ctx) && validate.errors) {
            throw CreateError(ajv.errorsText(validate.errors))
          } else return true
        }
      }
    }

    if (!this._config.name && this._config.run) {
      var match = this._config.run.toString().match(/function\s*(\w+)\s*\(/)

      if (match && match[1]) {
        this._config.name = match[1]
      } else {
        this._config.name = this._config.run.toString()
      }
    }
  }

  public get reportName() {
    return `STG:${this._config.name ? this._config.name : ''}`
  }

  public get name() {
    return this._config.name ?? ''
  }

  // может быть вызван как Promise
  // сделать все дубликаты и проверки методов для работы с промисами
  public execute(context: T): Promise<R | T>
  public execute(context: T, callback: CallbackFunction<R | T>): void
  public execute(
    err: Error | undefined,
    context: T,
    callback: CallbackFunction<R | T>,
  ): void
  public execute(
    _err: Error | T | undefined,
    _context?: T | CallbackFunction<R | T>,
    _callback?: CallbackFunction<R | T>,
  ): void | Promise<R | T> {
    // discover arguments
    let err: Error | undefined,
      context: T,
      __callback: CallbackFunction<R | T> | undefined

    if (arguments.length == 1) {
      context = _err as T
      //promise
    } else if (arguments.length == 2) {
      if (typeof _context == 'function') {
        // callback
        context = _err as T
        err = undefined
        __callback = _context as CallbackFunction<R | T>
      } else {
        //promise
        err = _err as Error
        context = _context as T
      }
    } else {
      // callback
      err = _err as Error
      context = _context as T
      __callback = _callback as CallbackFunction<R | T>
    }

    if (!this.run) {
      this.run = this.compile()
    }
    if (!__callback) {
      return new Promise((res, rej) => {
        this.execute(err, context as T, (err?: Error, ctx?: R | T) => {
          if (err) rej(err)
          else res(ctx as R | T)
        })
      })
    } else {
      const callback = __callback
      if (err && this._config.run && !can_fix_error(this._config.run)) {
        this.rescue(err, context, callback)
      } else {
        if (this.config.ensure) {
          this.ensure(this.config.ensure, context, (err_?: Error, ctx?: T) => {
            if (err || err_) {
              if (this._config.run && !can_fix_error(this._config.run)) {
                this.rescue(CreateError([err, err_]), ctx ?? context, callback)
              } else {
                // обработка ошибок может происходить внутри функции
                this.run?.(CreateError([err, err_]), ctx ?? context, callback)
              }
            } else {
              this.run?.(undefined, ctx ?? context, callback)
            }
          })
        } else if (this._config.validate) {
          this.validate(
            this._config.validate,
            context,
            (err_?: Error, ctx?: T) => {
              if (err || err_) {
                if (this._config.run && !can_fix_error(this._config.run)) {
                  this.rescue(
                    CreateError([err, err_]),
                    ctx ?? context,
                    callback,
                  )
                } else {
                  // обработка ошибок может происходить внутри функции
                  this.run?.(CreateError([err, err_]), ctx ?? context, callback)
                }
              } else {
                this.run?.(undefined, ctx ?? context, callback)
              }
            },
          )
        } else {
          this.run?.(undefined, context, callback)
        }
      }
    }
  }

  protected stage(
    err: Error | undefined,
    context: T,
    callback: CallbackFunction<T | R>,
  ) {
    if (this._config.run) {
      execute_callback(err, this._config.run, context, (err?: Error) => {
        if (err) {
          this.rescue(err, context, callback)
        } else {
          callback(undefined, context)
        }
      })
    } else {
      this.rescue(
        CreateError([this.reportName + ' reports: run is not a function', err]),
        context,
        callback,
      )
    }
  }

  public compile(rebuild: boolean = false): StageRun<T, R> {
    let res: StageRun<T, R>
    if (this.config.precompile) {
      this.config.precompile()
    }
    if (this._config.compile) {
      res = this._config.compile.call(this, rebuild)
    } else if (!this.run || rebuild) {
      res = this.stage
    } else {
      res = this.run
    }
    return res
  }
  // to be overridden by compile
  protected run?: StageRun<T, R>

  // объединение ошибок сделать
  protected rescue(
    err: Error | undefined,
    context: T,
    callback: CallbackFunction<T | R>,
  ) {
    if (err && !(err instanceof Error)) {
      if (typeof err == 'string') err = CreateError(err)
    }
    if (err && this._config.rescue) {
      execute_rescue(this._config.rescue, err, context, (_err?: Error) => {
        callback(_err, context)
      })
    } else {
      // отправить ошибку дальше
      callback(err, context)
    }
  }

  public toString() {
    return '[pipeline Stage]'
  }

  protected validate(
    validate: ValidateFunction<T>,
    context: T,
    callback: CallbackFunction<T>,
  ) {
    execute_validate(
      validate,
      context,
      (err: Error | undefined, result: boolean | undefined) => {
        if (err) {
          callback(err, context)
        } else {
          if (result) {
            if ('boolean' === typeof result) {
              callback(undefined, context)
            } else if (Array.isArray(result)) {
              callback(CreateError(result))
            } else {
              // ensure works
              callback(undefined, result)
            }
          } else {
            callback(CreateError(this.reportName + ' reports: T is invalid'))
          }
        }
      },
    )
  }
  protected ensure(
    ensure: EnsureFunction<T>,
    context: T,
    callback: CallbackFunction<T>,
  ) {
    execute_ensure(
      ensure,
      context,
      (err: Error | undefined, result: T | undefined) => {
        if (err) {
          callback(err, context)
        } else {
          if (result) {
            if ('boolean' === typeof result) {
              callback(undefined, context)
            } else if (Array.isArray(result)) {
              callback(CreateError(result))
            } else {
              // ensure works
              callback(undefined, result)
            }
          } else {
            callback(CreateError(this.reportName + ' reports: T is invalid'))
          }
        }
      },
    )
  }
}

export type EnsureParams<T, R> = {
  context: T
  callback: CallbackFunction<R> | undefined
  err: Error | undefined
  is_promise: boolean
}
