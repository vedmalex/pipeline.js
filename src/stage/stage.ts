import { z } from 'zod'

import { fromZodError } from 'zod-validation-error'
import { Context, OriginalObject } from './Context'
import { can_fix_error, ComplexError, CreateError } from './errors'
import { getStageConfig, isAnyStage, StageSymbol } from './getStageConfig'
import { StageConfig } from './StageConfig'
import { AllowedStage, AnyStage, CallbackFunction, isStageRun, Possible, RunPipelineFunction, StageRun } from './types'
import { execute_callback } from './utils/execute_callback'
import { execute_custom_run } from './utils/execute_custom_run'
import { execute_rescue } from './utils/execute_rescue'

export class Stage<R, C extends StageConfig<R> = StageConfig<R>> implements AnyStage<R> {
  public get config(): C {
    return this._config as C
  }
  [StageSymbol]: boolean
  protected _config!: C
  constructor()
  constructor(name: string)
  constructor(config: C)
  constructor(runFn: RunPipelineFunction<R>)
  constructor(stage: AnyStage<R>)
  constructor(config?: AllowedStage<R, C>) {
    this[StageSymbol] = true
    if (config) {
      if (typeof config === 'string') {
        this._config = { name: config } as C extends StageConfig<R> ? C : never
      } else {
        let res = getStageConfig<R, C>(config) as C extends StageConfig<R> ? C : never
        if (isAnyStage<R>(res)) {
          this._config = res.config as any
        } else {
          this._config = res
        }
      }
    } else {
      this._config = {} as C extends StageConfig<R> ? C : never
    }
  }

  public get reportName() {
    return `STG:${this._config.name ? this._config.name : ''}`
  }

  public toString() {
    return '[pipeline Stage]'
  }

  public get name() {
    return this._config.name ?? ''
  }

  // может быть вызван как Promise
  // сделать все дубликаты и проверки методов для работы с промисами
  public execute(context: R): Promise<R>
  public execute(context: R, callback: CallbackFunction<R>): void
  public execute(err: unknown, context: R, callback: CallbackFunction<R>): void
  public execute(
    _err?: unknown,
    _context?: unknown,
    _callback?: unknown,
  ): void | Promise<R> {
    // discover arguments
    let err: Possible<ComplexError>
    let not_ensured_context: R | R
    let __callback: Possible<CallbackFunction<R>> = undefined

    if (arguments.length == 1) {
      not_ensured_context = _err as R | R
      // promise
    } else if (arguments.length == 2) {
      if (typeof _context == 'function') {
        // callback
        not_ensured_context = _err as R | R
        err = undefined
        __callback = _context as CallbackFunction<R>
      } else {
        // promise
        err = _err as ComplexError
        not_ensured_context = _context as R | R
      }
    } else {
      // callback
      err = _err as ComplexError
      not_ensured_context = _context as R | R
      __callback = _callback as CallbackFunction<R>
    }

    if (!this.run) {
      this.run = this.compile()
    } else if (!this.config?.run) {
      // legacy run
      if (!isStageRun(this.run)) {
        var legacy = this.run

        this.run = execute_custom_run<R>(legacy as RunPipelineFunction<R>)
      }
    }

    const stageToRun = this.run?.bind(this)

    const input_is_context = Context.isProxy(not_ensured_context)
    const isProxible = input_is_context || (not_ensured_context && typeof not_ensured_context === 'object')

    let context = isProxible ? Context.ensure(not_ensured_context) : not_ensured_context
    if (isProxible) {
      context[OriginalObject] = true
    }
    // выполнить валидацию результата сразу же перед вызовом последнего значения
    if (!__callback) {
      return new Promise<R>((res, rej) => {
        this.execute(err, context, (err, ctx) => {
          if (err) {
            rej(err)
          } else {
            if (input_is_context) {
              res(ctx ?? context)
            } else {
              if (Context.isProxy(ctx)) {
                res(ctx.original as R)
              } else {
                res(ctx as unknown as R)
              }
            }
          }
        })
      })
    } else {
      const back: typeof __callback = (err, _ctx) => {
        if (input_is_context) {
          __callback?.(err, _ctx)
        } else {
          if (Context.isProxy<R>(_ctx)) {
            __callback?.(err, _ctx.original as R)
          } else {
            __callback?.(err, _ctx)
          }
        }
      }
      const successRescue = (ret: R) => {
        if (this._config.output) {
          this.validate(this._config.output, ret ?? context, (err_, ctx) => {
            if (err_) {
              this.rescue(err_, ctx ?? context, fail, successRescue)
            } else {
              back(err, ctx ?? context)
            }
          })
        } else {
          success(ret)
        }
      }
      const success = (ret: R) => back(undefined, ret ?? context)
      const fail = (err: unknown) => back(err, context)

      const callback = (err?, _ctx?: R) => {
        if (err) {
          this.rescue(err, _ctx ?? context, fail, successRescue)
        } else {
          if (this._config.output) {
            this.validate(this._config.output, _ctx ?? context, (err_, ctx) => {
              if (err_) {
                this.rescue(err_, ctx ?? context, fail, successRescue)
              } else {
                back(err, ctx ?? context)
              }
            })
          } else {
            back(err, _ctx ?? context)
          }
        }
      }

      if (err && this._config.run && !can_fix_error(this._config.run)) {
        this.rescue(err, context, fail, successRescue)
      } else {
        if (this._config.input) {
          this.validate(this._config.input, context, (err_, ctx) => {
            this.runStageMethod(err, err_, ctx, context, stageToRun, callback)
          })
        } else {
          stageToRun(undefined, context, callback)
        }
      }
    }
  }

  protected runStageMethod(
    err_: unknown,
    err: unknown,
    ctx: R | undefined,
    context: R,
    stageToRun: StageRun<R>,
    callback: CallbackFunction<R>,
  ) {
    if (err || err_) {
      if (this.config.run && !can_fix_error(this.config.run)) {
        this.rescue(CreateError([err, err_] as Array<Error>), ctx ?? context, callback, rescuedContext => {
          // ошибка обработана все хорошо, продолжаем
          stageToRun(undefined, rescuedContext, callback)
        })
      } else {
        // обработка ошибок может происходить внутри функции
        stageToRun(CreateError([err, err_] as Array<Error>), ctx ?? context, callback)
      }
    } else {
      stageToRun(undefined, ctx ?? context, callback)
    }
  }

  protected stage(err: unknown, context: R, callback: CallbackFunction<R>) {
    const back = callback
    const sucess = (ret: unknown) => back(undefined, (ret ?? context) as R)
    const fail = (err: unknown) => back(err, context)
    if (this._config.run) {
      if (context) {
        ;(execute_callback<R>).call(this, err, this._config.run, context, callback)
      } else {
        // возвращаем управление
        callback(err)
      }
    } else {
      const retErr: Array<any> = [this.reportName + ' reports: run is not a function']
      if (err) {
        retErr.push(err)
      }
      this.rescue(CreateError(retErr), context, fail, sucess)
    }
  }

  // to be overridden by compile
  protected run?: StageRun<R>
  /**
   * Compile the pipeline.
   * @param rebuild Whether to rebuild the pipeline.
   * @returns The compiled pipeline.
   */
  protected compile(rebuild: boolean = false): StageRun<R> {
    let res: StageRun<R>
    if (this.config.compile) {
      res = this.config.compile.call(this, rebuild)
    } else if (!this.run || rebuild) {
      res = this.stage
    } else {
      if (isStageRun<R>(this.run)) {
        res = this.run
      } else {
        res = execute_custom_run(this.run)
      }
    }
    return res
  }

  // объединение ошибок сделать
  // посмотреть что нужно сделать чтобы вызвать ошибку правильно!!!
  // в начале выполнения важен правильный callback, возможно без контекста
  // в конце важен и контекст ошибки? или не важен
  protected rescue(
    _err: unknown,
    context: R,
    fail: (err: unknown) => void,
    success: (ctx: R) => void,
  ) {
    let err: Possible<ComplexError>

    if (_err) {
      if (!(_err instanceof Error)) {
        err = CreateError(_err)
      } else {
        err = CreateError(_err)
      }
    } else {
      err = null
    }

    if (err && this._config.rescue) {
      execute_rescue(this._config.rescue, err, context, _err => {
        // здесь может быть исправлена ошибка, и контекст передается дальше на выполнение
        if (_err) {
          fail(_err)
        } else {
          success(context)
        }
      })
    } else {
      // отправить ошибку дальше
      // окончателная ошибка и выходим из выполнения
      if (err) {
        fail(err)
      } else {
        success(context)
      }
    }
  }

  protected rescue_async(
    _err: unknown,
    context: R,
  ): Promise<[unknown, R]> {
    return new Promise(resolve => {
      this.rescue(_err, context, err => {
        resolve([err, context])
      }, res => {
        resolve([undefined, res ?? context])
      })
    })
  }

  protected validate(
    validate: z.ZodType<R>,
    context: R,
    callback: CallbackFunction<R>,
  ) {
    validate.safeParseAsync(context)
      .then(result => {
        if (!result.success) {
          callback(CreateError(fromZodError(result?.error)), context)
        } else {
          callback(undefined, context)
        }
      })
  }
}
