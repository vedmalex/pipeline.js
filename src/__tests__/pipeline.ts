import 'jest'

import { Pipeline } from '../pipeline'
import { Stage } from '../stage'
import { ComplexError } from '../utils/ErrorList'

describe('Pipeline', function () {
  it('defaults', function (done) {
    var pipe = new Pipeline('defaultName')
    expect(pipe).toBeInstanceOf(Stage)

    expect('defaultName' === pipe.name).toBeTruthy()
    // expect(!pipe.config.stages).toEqual(false)
    expect(pipe.config.stages.length).toEqual(0)
    expect(!pipe.run).toEqual(true)

    expect(function () {
      pipe.compile()
    }).not.toThrow()
    pipe.execute({}, function (err, data) {
      expect(err == null).toBeTruthy()
      done()
    })
  })

  it('not allows to use constructor as a function', function (done) {
    try {
      var s = Pipeline()
    } catch (err) {
      done()
    }
  })

  it('addStage', function (done) {
    var pipe = new Pipeline()
    pipe.addStage(new Stage())
    expect(pipe.config.stages.length).toEqual(1)
    expect(pipe.config.stages[0] instanceof Stage).toEqual(true)
    done()
  })

  it('catch throw errors 1', function (done) {
    var pipe = new Pipeline()
    pipe.addStage(function (err, ctx, done) {
      throw new Error('error')
    })
    pipe.execute({}, function (err, ctx) {
      expect('error').toEqual(err?.message)
      done()
    })
  })

  it('catch throw errors 2', function (done) {
    var pipe = new Pipeline([
      function (err, ctx, done) {
        ctx.cnt = 1
        done(new Error('error'))
      },
      function (err, ctx, done) {
        ctx.cnt += 1
        done()
      },
    ])
    pipe.execute({}, function (err, ctx) {
      expect('error').toEqual(err?.message)
      done()
    })
  })

  it('catch throw errors 3', function (done) {
    var pipe = new Pipeline({
      rescue: function (err?) {
        if (err?.message == 'error') return false
      },
      stages: [
        function (err, ctx, done) {
          ctx.cnt = 1
          done()
        },
        function (err, ctx, done) {
          ctx.cnt += 1
          done(new Error('error'))
        },
      ],
    })
    pipe.execute({}, function (err, ctx) {
      expect(err).toBeUndefined()
      expect(ctx.cnt).toEqual(2)
      done()
    })
  })

  it('rescue works', function (done) {
    var pipe = new Pipeline()
    var st = new Stage({
      run: function (err, ctx, done) {
        throw new Error('error')
      },
      rescue: function (err, conext) {
        if (err.message !== 'error') return err
      },
    })
    pipe.addStage(st)
    pipe.execute({}, function (err, ctx) {
      expect(err == null).toBeTruthy()
      done()
    })
  })

  it('addStage converts function to Stage Instance', function (done) {
    var pipe = new Pipeline()
    pipe.addStage(function (err, ctx, done) {
      done()
    })
    expect(pipe.config.stages[0] instanceof Function).toBeTruthy()
    done()
  })

  it('addStage converts object to Stage Instance', function (done) {
    var pipe = new Pipeline()
    pipe.addStage({
      run: function (err, ctx, done) {
        done()
      },
    })
    expect(pipe.config.stages[0] instanceof Stage).toBeTruthy()
    done()
  })

  it('accept array of stages', function (done) {
    var f1 = function (err, ctx, done) {
      done()
    }
    var f2 = function (err, ctx, done) {
      done()
    }

    var pipe = new Pipeline([f1, f2])
    pipe.addStage(function (err, ctx, done) {
      done()
    })
    expect(pipe.config.stages[0] instanceof Function).toBeTruthy()
    expect(pipe.config.stages[1] instanceof Function).toBeTruthy()

    done()
  })

  it('accept empty addStages', function (done) {
    var pipe = new Pipeline()
    ;(pipe as any).addStage()
    expect(pipe.config.stages.length).toEqual(0)
    done()
  })

  it('compile', function (done) {
    var pipe = new Pipeline()
    pipe.addStage(new Stage())
    pipe.compile()
    expect(typeof pipe.run == 'function').toBeTruthy()
    pipe.addStage(new Stage())
    expect(!(pipe as any).run).toBeTruthy()
    done()
  })

  it('execute must call compile and ensure', function (done) {
    let ensure = 0
    let compile = 0

    let pipe = new Pipeline({
      precompile: () => {
        compile++
      },
      ensure: function (ctx, callback) {
        ensure++
      },
    } as any)
    pipe.addStage(
      new Stage(function (err, context, done) {
        done()
      }),
    )

    pipe.execute({})
    expect(ensure).toEqual(1)
    expect(compile).toEqual(1)

    pipe.execute({})
    expect(ensure).toEqual(2)
    expect(compile).toEqual(1)
    done()
  })

  it('executes pipes in pipes', function (done) {
    var pipe = new Pipeline()
    var nestedpipe = new Pipeline()
    nestedpipe.addStage(function (err, ctx, done) {
      ctx.item = 1
      done()
    })
    pipe.addStage(nestedpipe)
    pipe.execute(
      {
        item: 0,
      },
      function (err, ctx) {
        expect(err).toBeUndefined()
        expect(1).toEqual(ctx.item)
        done()
      },
    )
  })

  it('context catch all errors', function (done) {
    var pipe = new Pipeline()
    var ctx1 = {
      s1: false,
      s2: false,
      s3: false,
    }
    var error = new Error('THE ERROR')

    var s1 = new Stage(function (err, context, done) {
      expect(context).toEqual(ctx1)
      context.s1 = true
      done()
    })
    var s2 = new Stage(function (err, context, done) {
      expect(context).toEqual(ctx1)
      context.s2 = true
      done(error)
    })
    var s3 = new Stage(function (err, context, done) {
      expect(true).toEqual(false)
      context.s3 = true
      done()
    })

    pipe.addStage(s1)
    pipe.addStage(s2)

    pipe.execute(ctx1, function (err, ctx) {
      expect(err).toEqual(error)
      // expect(ctx1.hasErrors()).toEqual(true);
      // expect(ctx1.getErrors()[0] == error).toEqual(true);
      expect(ctx1.s1).toEqual(true)
      expect(ctx1.s2).toEqual(true)
      expect(ctx1.s3).toEqual(false)
      done()
    })
  })

  it('ensure Context Error use', function (done) {
    var ctx = {}
    ctx.SomeValue = 1
    var pipe = new Pipeline()
    var error = new Error('context not ready')
    var stage1 = {
      run: function (err, context, done) {
        context.SomeValue += 1
        done()
      },
      ensure: function (context, callback) {
        if (context.SomeValue !== 1) callback(error)
        else callback(null, context)
      },
    }
    pipe.addStage(new Stage(stage1))
    var stage2 = {
      ensure: function WV(context, callback) {
        if (context.SomeValue !== 1) {
          callback(new Error(/* this.reportName +  */ ': Wrong Value'))
        } else callback(null, context)
      },
      run: undefined, // so it will be 0
    }
    var s2 = new Stage(stage2)
    pipe.addStage(s2)
    pipe.execute(ctx, function (err) {
      expect(err).not.toBeUndefined()
      expect(
        /Error: STG: reports: run is not a function/.test(
          err.errors[0].toString(),
        ),
      ).toBeTruthy()
      done()
    })
  })

  it('can do subclassing of Pipeline', function (done) {
    class newPipe extends Pipeline {
      constructor() {
        super()
        this.addStage(new Stage())
        this.addStage(new Stage())
        this.addStage(new Stage())
      }
    }

    var p1 = new newPipe()
    expect(p1.config.stages.length).toEqual(3)
    var p2 = new newPipe()
    p2.addStage(new Stage())
    expect(p2.config.stages.length).toEqual(4)
    var p3 = new newPipe()
    expect(p3.config.stages.length).toEqual(3)
    done()
  })

  it('allow reenterability', function (done) {
    debugger
    var pipe = new Pipeline()

    pipe.addStage(function (context, done) {
      process.nextTick(function () {
        context.one++
        done()
      })
    })

    pipe.addStage(function (context, done) {
      process.nextTick(function () {
        context.one += 1
        done()
      })
    })

    pipe.addStage(function (context, done) {
      context.one += 5
      done()
    })

    var l = 0

    function gotit() {
      if (++l == 10) done()
    }
    for (var i = 0; i < 10; i++) {
      ;(function () {
        var ctx1 = {
          one: 1,
        }
        pipe.execute(ctx1, function (err, data) {
          expect(data.one).toEqual(8)
          gotit()
        })
      })()
    }
  })
})
