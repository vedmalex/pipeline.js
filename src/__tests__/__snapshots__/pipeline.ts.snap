// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline ensure Context Error use 1`] = `[Error: Complex Error]`;

exports[`Pipeline ensure Context Error use 2`] = `[Error: done call more than once]`;

exports[`Pipeline ensure Context Error use 3`] = `[Error: done call more than once]`;

exports[`Pipeline ensure Context Error use 4`] = `[Error: done call more than once]`;
